File: src/main/java/com/polygloat/model/Source.java
package com.polygloat.model;

import com.polygloat.dtos.PathDTO;
import lombok.*;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

@EqualsAndHashCode(exclude = {"translations", "repository"}, callSuper = true)
@Entity
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"repository_id", "name"}, name = "source_repository_id_name"),
})
public class Source extends AuditModel {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    //@Analyzer(impl = Analyzer.class)
    private String name;

    @ManyToOne
    private Repository repository;

    @Builder.Default
    @OneToMany(mappedBy = "source")
    private Set<Translation> translations = new HashSet<>();

    public Optional<Translation> getTranslation(String abbr) {
        return this.getTranslations().stream().filter(t -> t.getLanguage().getAbbreviation().equals(abbr)).findFirst();
    }

    public PathDTO getPath() {
        return PathDTO.fromFullPath(this.getName());
    }

}

File: src/main/kotlin/io/polygloat/model/Source.kt
package io.polygloat.model

import com.unboundid.util.Mutable
import io.polygloat.dtos.PathDTO
import java.util.*
import javax.persistence.*
import kotlin.collections.HashSet

@Entity
@Table(uniqueConstraints = [UniqueConstraint(columnNames = ["repository_id", "name"], name = "source_repository_id_name")])
//Todo: Rename to Key
data class Source(
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long? = null,

        var name: String? = null,
) : AuditModel() {
    @ManyToOne
    var repository: Repository? = null

    @OneToMany(mappedBy = "key")
    var translations: MutableSet<Translation> = HashSet()

    constructor(id: Long? = null,
                name: String? = null,
                repository: Repository? = null,
                translations: MutableSet<Translation> = HashSet()
    ) : this(id, name) {
        this.repository = repository
        this.translations = translations
    }

    fun getTranslation(abbr: String): Optional<Translation> {
        return translations.stream().filter { t: Translation -> t.language.abbreviation == abbr }.findFirst()
    }

    val path: PathDTO
        get() = PathDTO.fromFullPath(name)
}